### PR TITLE
feat!: type inference for evaluation types

### LINKED ISSUE #8485: [Types] Improve types for `evaluate` and `evaluateHandle` on various types (and `JSHandle<T>`)
Currently the `evaluate` and `evaluateHandle` functions and `JSHandle<T>` itself are very loosely typed, for example, this is allowed:

```ts
async function demo(handle1: JSHandle<number>, handle2: JSHandle<string>) {
	const notActuallyBoolean: JSHandle<'wait what'> = await handle1.evaluateHandle<JSHandle<boolean>>(
		  (num, notActuallyNumber: number, notActuallyString: string) => 'hi', handle2, 42);
}
```

**Problems are:**
- Argument types are not checked (except for the subject handle itself).
- Return type is not checked for `evaluateHandle`.
- `JSHandle<A> extends JSHandle<B>` for any `A`,`B`. Same for `ElementHandle<E>`.

# Solution

This could be solved by typing the functions as follows:

```ts
type Passed<H extends SerializableOrJSHandle> = H extends JSHandle<infer U> ? U : H;
type PassedArgs<Args extends SerializableOrJSHandle[]> = { [K in keyof Args]: Passed<Args[K]> };

class JSHandle<HandleObjectType> {
  /** Make sure that not e.g. `JSHandle<string> extends `JSHandle<number>` */
  private _dummy?: HandleObjectType;

  evaluateHandle<Args extends SerializableOrJSHandle[], Return>(
      pageFunction: (subject: HandleObjectType, ...args: PassedArgs<Args>) => Return,
      ...args: Args
  ): Promise<JSHandle<Return>> { ... }

  evaluate<
      Args extends SerializableOrJSHandle[],
      Return extends Serializable | PromiseLike<Serializable> | void
  >(
      pageFunction: (subject: HandleObjectType, ...args: PassedArgs<Args>) => Return,
      ...args: Args
  ): Promise<Awaited<Return>> { ... }

  ...
}

class Frame {
  evaluateHandle<Args extends SerializableOrJSHandle[], Return>(
      pageFunction: (...args: PassedArgs<Args>) => Return,
      ...args: Args
  ): Promise<JSHandle<Return>> { ... }

  evaluate<
      Args extends SerializableOrJSHandle[],
      Return extends Serializable | PromiseLike<Serializable> | void
  >(
      pageFunction: (...args: PassedArgs<Args>) => Return,
      ...args: Args
  ): Promise<Awaited<Return>> { ... }

  ...
}

// Same idea for ExecutionContext, Page, DOMWorld
```

With this version, the TypeScript compiler will complain about the `demo` function from earlier.

Feel free to copy parts of this code if you would want to.

Note that without the `_dummy` field it is not just the case that we can assign any `JSHandle` to any other, but we also have that `Passed<ElementHandle<HTMLElement>> extends unknown` instead of `HTMLElement`.

# Wrapper/workaround for current Puppeteer versions (informational)

You can already accomplish this in current versions using the following wrapper functions:

```ts
// Workaround: Need the ElementHandle case because inferring U on JSHandle gives unknown because currently JSHandle<A> extends JSHandle<B> for any A,B...
type HandleValue<H extends JSHandle> = H extends ElementHandle<infer U> ? U
	  : H extends JSHandle<infer U> ? U : never;

type Passed<H extends SerializableOrJSHandle> = H extends JSHandle ? HandleValue<H> : H;
type PassedArgs<Args extends SerializableOrJSHandle[]> = { [K in keyof Args]: Passed<Args[K]> };

type PageFunction<Target extends Frame | Page | ExecutionContext | DOMWorld | JSHandle,
	  Args extends SerializableOrJSHandle[], Return> =
	  Target extends JSHandle
			? (obj: HandleValue<Target>, ...args: PassedArgs<Args>) => Return
			: (...args: PassedArgs<Args>) => Return;

/** Typed version of {@link Frame#evaluateHandle}, {@link JSHandle#evaluateHandle}, etc. */
export async function evaluateHandle<Subject extends Frame | Page | ExecutionContext | DOMWorld | JSHandle,
	  Args extends SerializableOrJSHandle[], Return>(
	  subject: Subject, pageFunction: PageFunction<Subject, Args, Return>, ...args: Args): Promise<JSHandle<Return>> {
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore This compiles for all Targets individually
	return await subject.evaluateHandle(pageFunction, ...args) as JSHandle<Return>;
}

/** Typed version of {@link Frame#evaluate}, {@link JSHandle#evaluate}, etc. */
export async function evaluate<Subject extends Frame | Page | ExecutionContext | DOMWorld | JSHandle,
	  Args extends SerializableOrJSHandle[], Return extends Serializable | PromiseLike<Serializable> | void>(
	  subject: Subject, pageFunction: PageFunction<Subject, Args, Return>, ...args: Args): Promise<Awaited<Return>> {
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore This compiles for all Targets individually
	return await subject.evaluate(pageFunction as never, ...args) as Awaited<Return>;
}
```

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR greatly improves the types within Puppeteer:

- **Almost everything** is auto-deduced.
  - Parameters don't need to be specified in the function. They are deduced from the spread.
  - Return types don't need to be specified. They are deduced from the function. (More on this below)
  - Selections based on tag names correctly deduce element type, similar to TypeScript's mechanism for `getElementByTagName`.
- [**BREAKING CHANGE**] We've removed the ability to declare return types in type arguments for the following reasons:
  1. Setting them will indubitably break auto-deduction.
  2. You can just use `as ...` in TypeScript to coerce the correct type (given it makes sense).
- [**BREAKING CHANGE**] `waitFor` is officially gone.

To migrate to these changes, there are only four things you may need to change:
- If you set a return type using the `ReturnType` type parameter, remove it and use `as ...` and `HandleFor` (if necessary).
⛔ `evaluate<ReturnType>(a: number, b: number) => {...}, a, b)`
✅ `(await evaluate(a, b) => {...}, a, b)) as ReturnType`
⛔ `evaluateHandle<ReturnType>(a: number, b: number) => {...}, a, b)`
✅ `(await evaluateHandle(a, b) => {...}, a, b)) as HandleFor<ReturnType>`
- If you set any type parameters in the *parameters* of an evaluation function, remove them.  
⛔ `evaluate(a: number, b: number) => {...}, a, b)`
✅ `evaluate(a, b) => {...}, a, b)`
- If you set any type parameters in the method's declaration, remove them.
⛔ `evaluate<(a: number, b: number) => void>((a, b) => {...}, a, b)`
✅ `evaluate(a, b) => {...}, a, b)`

Fixes: #8485

### CODE REVIEW

**Comment by stevenwdv** [comment]
You could also consider changing `ElementHandle<T extends Element>` to `ElementHandle<T extends ParentNode>` to officially support `Document` and `ShadowRoot` as well; see my PR #8540

**Comment by jrandolf-google** [comment]
@stevenwdv There are plans to handle `Node`s! Stay tuned.

File: test/src/waittask.spec.ts [hunk 1]
```diff
@@ -29,122 +28,6 @@ describe('waittask specs', function () {
   setupTestBrowserHooks();
   setupTestPageAndContextHooks();
 
-  describe('Page.waitFor', function () {
```
Comment (OrKoN) [inline test/src/waittask.spec.ts:32]: I think some of the tests don't use the deprecated waitFor? I think we should keep them (I wonder if the coverage test would complain)

Comment (jrandolf) [reply test/src/waittask.spec.ts:32]: It's okay because we tests these functions in other places.

**Comment by OrKoN** [comment]
let's document the breaking part of this PR better? let's perhaps include an example of what worked before and how it should be handled now. E.g., before `handle.evaluate<MyCustomType>(() => ..)`, after `handle.evaluate(() => ..) as MyCustomType` if I understood it right?

**Comment by OrKoN** [comment]
By documenting I mean let's include an example into the PR description (and also copy everything from the PR description into the commit message when landing this).

**Comment by OrKoN** [comment]
nit: let's rephrase the title to be more specific, e.g., `implement type inference for evaluate types`

**Review by OrKoN — APPROVED** [review]

**Comment by karlhorky** [comment]
@jrandolf Did the following lines introduce bugs in the types?

According to [the docs](https://pptr.dev/#?product=Puppeteer&version=v15.1.1&show=api-pageevalselector-pagefunction-args-1) a `pageFunction` should only be a function, but here a `string` is also being allowed? 

- https://github.com/puppeteer/puppeteer/pull/8547/files#diff-faca1aee95cf49cbbf62a8fbad766f4d6f943b9a830466c962b7066909a4e54aR184
- https://github.com/puppeteer/puppeteer/pull/8547/files#diff-faca1aee95cf49cbbf62a8fbad766f4d6f943b9a830466c962b7066909a4e54aR195

**Comment by jrandolf-google** [comment]
> @jrandolf Did the following lines introduce bugs in the types?
> 
> According to [the docs](https://pptr.dev/#?product=Puppeteer&version=v15.1.1&show=api-pageevalselector-pagefunction-args-1) a `pageFunction` should only be a function, but here a `string` is also being allowed?
> 
> * https://github.com/puppeteer/puppeteer/pull/8547/files#diff-faca1aee95cf49cbbf62a8fbad766f4d6f943b9a830466c962b7066909a4e54aR184
> * https://github.com/puppeteer/puppeteer/pull/8547/files#diff-faca1aee95cf49cbbf62a8fbad766f4d6f943b9a830466c962b7066909a4e54aR195

Thanks for the comment. The documentation is out of date (we removed the current generator as it cannot support our new types). In any case, it was always okay to use a string actually. Much of this PR attempts to amend the problems of our Puppeteer forefathers!

**Comment by karlhorky** [comment]
Ok, thanks for that. A bit confusing currently unfortunately! (mismatching types / docs). But maybe this will be remedied fairly quickly :)

**Pull request merged**